File: lab-1-rcopy-master/java/RCopy.java
import java.io.*;
public class RCopy {
    public static void main (String[] args) {
        try {
            int BUFFER_SIZE=1024;
            byte[] buffer = new byte [BUFFER_SIZE];
            int size=0;

            RandomAccessFile inFile = new RandomAccessFile(args[0], "r");
            FileOutputStream outStream = new FileOutputStream(args[1]);
           
            if(inFile.length()>BUFFER_SIZE){
            	size++;
            	inFile.seek(inFile.length()-BUFFER_SIZE*size);
            	
            }else{
            	size++;
            }
	        int read=inFile.read (buffer);
            while(read!=-1){
            	int start=0,end=read-1;
                byte Data;
                while(start<end) {
                	Data=buffer[start];
                	buffer[start]=buffer[end];
                	buffer[end]=Data;
                	start++;
                	end--;
                }
                outStream.write (buffer, 0, read);
                if((inFile.length()-BUFFER_SIZE*size)>BUFFER_SIZE){
                	size++;
                	inFile.seek(inFile.length()-BUFFER_SIZE*size);
                    read=inFile.read (buffer);
                }else if((inFile.length()-BUFFER_SIZE*size)>0){
                	inFile.seek(0);
                    read=inFile.read(buffer, 0, (int) (inFile.length()-BUFFER_SIZE*size));
                	size++;

                }else
                {
                	break;
                }
                
            }
            
            inFile.close ();
            outStream.close ();
        } catch (IOException e) {
            System.out.println ("IO-Exception while: " + e.toString());
        }
    }
}

File: lab-1-rcopy-master/cpp/rcopy.cc
#include <iostream>
#include <fstream>
#define BUFFER_SIZE 1024
using namespace std;

int main(const int argc, const char **argv)
{
    char buffer[BUFFER_SIZE];
    char Data;
    int size=0;

    ifstream inputFile;
    inputFile.open(argv[1]);
    ofstream outFile;
    outFile.open(argv[2]);

    int start = inputFile.tellg();
    inputFile.seekg (0, ios::end);
    int end = inputFile.tellg();
    inputFile.seekg (0, ios::beg);
    int length =end-start;

    
    if(length>BUFFER_SIZE){
        size++;
        inputFile.seekg(length-size*BUFFER_SIZE,ios::beg);
    }else{
        size++;
    }
    int read =inputFile.read(buffer,BUFFER_SIZE).gcount();

    while(read!=-1) {
        int end=read-1;
        int beg=0;
        while(beg<end){
            Data=buffer[beg];
            buffer[beg]=buffer[end];
            buffer[end]=Data;
            beg++;
            end--;
        }
        outFile.write(buffer,read);
        if((length-size*BUFFER_SIZE)>BUFFER_SIZE){
            size++;
            inputFile.seekg(length-size*BUFFER_SIZE,ios::beg);
            read=inputFile.read(buffer,BUFFER_SIZE).gcount();
        }else if ((length-size*BUFFER_SIZE)>0){
            inputFile.seekg(0,ios::beg);
            read=inputFile.read(buffer,length-size*BUFFER_SIZE).gcount();
            size++;
        }else{
            break;
        }
    }
    inputFile.close();
    outFile.close();
    return 0;
}

